In [13]:
import pandas as pd
from utills import clean_names
from utills import get_bool_name

In [14]:
trues = []   #nombre con coincidencias
falses = []  # nombre sin coincidencias
m_igual = []

In [15]:
G = open("prob75.txt", "r")
linea = G.readline()
while len(linea)!=0:
    part = linea.strip().split(" | ")
    x, y, a, b = part
    TF = get_bool_name(x, y)
    if float(b)>0.91:
        m_igual.append([x, y, a, b])
    elif TF==True:
        trues.append([x, y, a, b])
    else:
        falses.append([x, y, a, b])
    linea = G.readline()
G.close()

In [16]:
print(len(trues), len(m_igual))

994 340


In [17]:
trues2 = []
m_igual2 = []
for i in trues:
    if i not in trues2:
        trues2.append(i)
for i in m_igual:
    if i not in m_igual2:
        m_igual2.append(i)

In [18]:
print(len(trues2), len(m_igual2))

989 340


In [19]:
veridicos = [i[0] for i in m_igual2]

In [20]:
trues3 = [i for i in trues2 if i[0] not in veridicos]

In [21]:
veridicos2 = m_igual2 + trues3
len(veridicos2)

1317

In [23]:
veridicos2.sort()

In [22]:
veridicos2 + ['juan carlos cordoba', 'juan cordoba', 'pco53', '0.77419',
'hernán ruffo', 'hernan m. ruffo', 'pru227', '0.81481',
'michael d. könig', 'michael king', 'pki400', '0.78571'
'miguel a león-ledesma', 'miguel leon-ledesma', 'ple108', '0.90000',
'renée b. adams', 'renee adams', 'pad158', '0.80000']


[['a. kerem co?ar', 'a. kerem cosar', 'pco958', '0.92857'],
 ['abhijit v. banerjee', 'abhijit banerjee', 'pba1710', '0.91429'],
 ['achyuta adhvaryu', 'achyuta r. adhvaryu', 'pad77', '0.91429'],
 ['adam m. lavecchia', 'adam m lavecchia', 'pla823', '0.96970'],
 ['adriaan r. soetevent', 'adriaan soetevent', 'pso65', '0.91892'],
 ['adriano a rampini', 'adriano a. rampini', 'pra423', '0.97143'],
 ['adrienne m. lucas', 'adrienne m lucas', 'plu223', '0.96970'],
 ["ahuva mu'alem", "ahuva mu'alem", 'pmu575', '1.00000'],
 ['aico van vuuren', 'aico van vuuren', 'pva4', '1.00000'],
 ['alain de janvry', 'alain de janvry', 'pde613', '1.00000'],
 ['alan m taylor', 'alan m. taylor', 'pta46', '0.96296'],
 ['alan de bromhead', 'alan de bromhead', 'pde717', '1.00000'],
 ['alexander d. rothenberg', 'alexander d rothenberg', 'pro1171', '0.97778'],
 ['alexander gelber', 'alexander m. gelber', 'pge116', '0.91429'],
 ['alexander karaivanov', 'alexander k. karaivanov', 'pka38', '0.93023'],
 ['alfredo di tillio

In [31]:
veridicos2 = [[clean_names(veridicos2[i][0]), veridicos2[i][1], veridicos2[i][2]] for i in range(len(veridicos2))]
for i in range(len(veridicos2)):
    print(i, veridicos2[i])

0 ['a mushfiq mobarak', 'ahmed mushfiq mobarak', 'pmo232']
1 ['a. kerem co?ar', 'a. kerem cosar', 'pco958']
2 ['a. mushfiq mobarak', 'ahmed mushfiq mobarak', 'pmo232']
3 ['aaron s. edlin', 'aaron edlin', 'ped12']
4 ['aaron swoboda', 'aaron m. swoboda', 'psw23']
5 ['abel winn', 'abel m. winn', 'pwi202']
6 ['abhijit v. banerjee', 'abhijit banerjee', 'pba1710']
7 ['achyuta adhvaryu', 'achyuta r. adhvaryu', 'pad77']
8 ['adam ashcraft', 'adam b. ashcraft', 'pas35']
9 ['adam b. jaffe', 'adam jaffe', 'pja49']
10 ['adam copeland', 'adam matthew copeland', 'pco241']
11 ['adam m. guren', 'adam guren', 'pgu538']
12 ['adam m. lavecchia', 'adam m lavecchia', 'pla823']
13 ['adam m. rosen', 'adam michael rosen', 'pro441']
14 ['adnan q. khan', 'adnan khan', 'pkh414']
15 ['adnan q. khan', 'adnan khan', 'pkh420']
16 ['adriaan r. soetevent', 'adriaan soetevent', 'pso65']
17 ['adriano a rampini', 'adriano a. rampini', 'pra423']
18 ['adrien vigier', 'adrien henri vigier', 'pvi322']
19 ['adrienne m. lucas',

887 ['michael d. bordo', 'michael david bordo', 'pbo243']
888 ['michael d. makowsky', 'michael douglas makowsky', 'pma921']
889 ['michael d. whinston', 'michael whinston', 'pwh46']
890 ['michael g. palumbo', 'michael g. plummer', 'ppl25']
891 ['michael g. palumbo', 'michael palumbo', 'ppa1156']
892 ['michael geruso', 'michael l. geruso', 'pge198']
893 ['michael j. dickstein', 'michael j. dueker', 'pdu108']
894 ['michael j. fishman', 'michael j. fleming', 'pfl35']
895 ['michael j. hilmer', 'michael j. hicks', 'phi52']
896 ['michael j. keen', 'michael keen', 'pke22']
897 ['michael j. keen', 'michael kleemann', 'pkl81']
898 ['michael j. meurer', 'michael j. murray', 'pmu340']
899 ['michael j. roberts', 'mark j. roberts', 'pro190']
900 ['michael j. roberts', 'michael james roberts', 'pro207']
901 ['michael keane', 'michael p keane', 'pke18']
902 ['michael l. lemmon', 'michael l. lahr', 'pla105']
903 ['michael m. ting', 'michael m. tamminga', 'pta793']
904 ['michael mcmahon', 'michael franc

In [30]:
veridicos2[680]=['juan pablo torres-martinez', 'juan pablo torres-martinez', 'pto20']
veridicos2[674]=['juan f rubio-ramirez', 'juan f rubio-ramirez', 'pru25']

In [52]:
len(veridicos2)

1317

In [46]:
mini_df = pd.DataFrame(veridicos2, columns=["author", "name", "Short-Id"])
mini_df

author                       name Short-Id
0          a mushfiq mobarak      ahmed mushfiq mobarak   pmo232
1             a. kerem co?ar             a. kerem cosar   pco958
2         a. mushfiq mobarak      ahmed mushfiq mobarak   pmo232
3             aaron s. edlin                aaron edlin    ped12
4              aaron swoboda           aaron m. swoboda    psw23
...                      ...                        ...      ...
1312          yishay d. maoz          yishay david maoz   pma463
1313           young han lee             young hoon lee   ple304
1314           young han lee             yung-chuan lee   ple984
1315         yukiko u. saito         yukiko umeno saito  psa1086
1316  zarek c. brot-goldberg  zarek chase brot-goldberg   pbr639

[1317 rows x 3 columns]

### Un total de 1317 autores con coincidencias
#### Aun asi, puede haber  un autor repetido por un determinado rango de error

### Se procede a merge-ar los DataFrames

In [53]:
import pandas as pd
import numpy as np
import os

In [54]:
df_eco= pd.read_csv('emetrp.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)
df_aer= pd.read_csv('aecrev_v2.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)
df_jpe= pd.read_csv('jpolec.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)
df_qje= pd.read_csv('qjecon.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)
df_res= pd.read_csv('restud.csv', delimiter=';',encoding = "ISO-8859-1", error_bad_lines=False)

In [55]:
df=df_aer.append(df_eco,ignore_index=True)
df=df.append(df_qje,ignore_index=True)
df=df.append(df_jpe,ignore_index=True)
df=df.append(df_res,ignore_index=True)

In [56]:
df.drop(columns=['handle','title','vol','issue','page','end page'], inplace=True)

In [57]:
df

journal    year                   author
0        American Economic Review  2000.0          D. Gale Johnson
1        American Economic Review  2000.0        Dmitriy Stolyarov
2        American Economic Review  2000.0          Boyan Jovanovic
3        American Economic Review  2000.0        Eric S. Rosengren
4        American Economic Review  2000.0                 Joe Peek
...                           ...     ...                      ...
12935  Review of Economic Studies  2019.0      Jeong Ho (John) Kim
12936  Review of Economic Studies  2019.0          Eduardo Morales
12937  Review of Economic Studies  2019.0              Gloria Sheu
12938  Review of Economic Studies  2019.0            Andrés Zahler
12939  Review of Economic Studies  2019.0  Eduardo Souza-Rodrigues

[12940 rows x 3 columns]

In [58]:
df_repec=pd.read_csv('autores_repec.csv', delimiter=',')
df_repec

Name Short-Id
0              Arild Aakvik     paa1
1                 Eline Aas    paa10
2           Arnstein Aassve    paa11
3      Susan Ariel Aaronson    paa12
4           Daniel Aaronson    paa13
...                     ...      ...
70758   Eduardo Zylberstajn     pzy5
70759     Adam Zylbersztejn     pzy6
70760          Robert Zymek     pzy7
70761      Yanos Zylberberg     pzy8
70762       Benjamin Zycher     pzy9

[70763 rows x 2 columns]

In [59]:
df_final = pd.merge(df, df_repec, how = 'left',left_on ='author', right_on = 'Name')
df_final

journal    year                   author  \
0        American Economic Review  2000.0          D. Gale Johnson   
1        American Economic Review  2000.0        Dmitriy Stolyarov   
2        American Economic Review  2000.0          Boyan Jovanovic   
3        American Economic Review  2000.0        Eric S. Rosengren   
4        American Economic Review  2000.0                 Joe Peek   
...                           ...     ...                      ...   
13025  Review of Economic Studies  2019.0      Jeong Ho (John) Kim   
13026  Review of Economic Studies  2019.0          Eduardo Morales   
13027  Review of Economic Studies  2019.0              Gloria Sheu   
13028  Review of Economic Studies  2019.0            Andrés Zahler   
13029  Review of Economic Studies  2019.0  Eduardo Souza-Rodrigues   

                  Name Short-Id  
0      D. Gale Johnson   pjo254  
1                  NaN      NaN  
2      Boyan Jovanovic    pjo20  
3                  NaN      NaN  
4             Joe Peek    ppe90  
...                ...      ...  
13025              NaN      NaN  
13026              NaN      NaN  
13027      Gloria Sheu   psh447  
13028              NaN      NaN  
13029              NaN      NaN  

[13030 rows x 5 columns]

### Primer DataFrame creado con las coincidencias
##### No se requirio ningun filtro

In [60]:
df_final_1 = df_final.dropna()
df_final_1

journal    year             author  \
0        American Economic Review  2000.0    D. Gale Johnson   
2        American Economic Review  2000.0    Boyan Jovanovic   
4        American Economic Review  2000.0           Joe Peek   
5        American Economic Review  2000.0    Larry Samuelson   
6        American Economic Review  2000.0  George J. Mailath   
...                           ...     ...                ...   
13019  Review of Economic Studies  2019.0     Elisa Faraglia   
13020  Review of Economic Studies  2019.0      Albert Marcet   
13021  Review of Economic Studies  2019.0    Rigas Oikonomou   
13022  Review of Economic Studies  2019.0       Andrew Scott   
13027  Review of Economic Studies  2019.0        Gloria Sheu   

                    Name Short-Id  
0        D. Gale Johnson   pjo254  
2        Boyan Jovanovic    pjo20  
4               Joe Peek    ppe90  
5        Larry Samuelson    psa80  
6      George J. Mailath     pma6  
...                  ...      ...  
13019     Elisa Faraglia   pfa164  
13020      Albert Marcet   pma173  
13021    Rigas Oikonomou     poi7  
13022       Andrew Scott    psc65  
13027        Gloria Sheu   psh447  

[7458 rows x 5 columns]

### Segundo Dataframe con valores de repec no encontrados
#### Se procede a procesar usando un filtro de comparacion

In [61]:
df_final_0 = df_final[df_final.isna().any(axis=1)]
df_final_0 = df_final_0[['journal', 'year', 'author']]
df_final_0.sort_values(by='author')

journal    year                author
3027     American Economic Review  2013.0       ?kos Valentinyi
12908  Review of Economic Studies  2019.0     A Mushfiq Mobarak
2119     American Economic Review  2010.0  A. Joshua Strickland
3739     American Economic Review  2016.0        A. Kerem Co?ar
12341  Review of Economic Studies  2014.0             A. Norets
...                           ...     ...                   ...
11797                         NaN     NaN                   NaN
11798                         NaN     NaN                   NaN
12578                         NaN     NaN                   NaN
12579                         NaN     NaN                   NaN
12580                         NaN     NaN                   NaN

[5572 rows x 3 columns]

In [62]:
autors_ = list(df_final_0["author"])
lista = []
for i in autors_:
    if type(i)==float:
        lista.append(i)
    else:
        lista.append(i.lower())
df_final_0['author'] = lista

In [63]:
df_final_0

journal    year                   author
1        American Economic Review  2000.0        dmitriy stolyarov
3        American Economic Review  2000.0        eric s. rosengren
7        American Economic Review  2000.0             avner shaked
8        American Economic Review  2000.0        david e. wildasin
10       American Economic Review  2000.0        thomas j. nechyba
...                           ...     ...                      ...
13024  Review of Economic Studies  2019.0            junehyuk jung
13025  Review of Economic Studies  2019.0      jeong ho (john) kim
13026  Review of Economic Studies  2019.0          eduardo morales
13028  Review of Economic Studies  2019.0            andrés zahler
13029  Review of Economic Studies  2019.0  eduardo souza-rodrigues

[5572 rows x 3 columns]

### Finalmente se obtiene el Dataframe filtrado
#### Aquellos autores con `NaN` en la columna `Name` no obtuvieron coincidencias

In [66]:
mini_df

author                       name Short-Id
0          a mushfiq mobarak      ahmed mushfiq mobarak   pmo232
1             a. kerem co?ar             a. kerem cosar   pco958
2         a. mushfiq mobarak      ahmed mushfiq mobarak   pmo232
3             aaron s. edlin                aaron edlin    ped12
4              aaron swoboda           aaron m. swoboda    psw23
...                      ...                        ...      ...
1312          yishay d. maoz          yishay david maoz   pma463
1313           young han lee             young hoon lee   ple304
1314           young han lee             yung-chuan lee   ple984
1315         yukiko u. saito         yukiko umeno saito  psa1086
1316  zarek c. brot-goldberg  zarek chase brot-goldberg   pbr639

[1317 rows x 3 columns]

In [67]:
df_final_0_1 = pd.merge(df_final_0, mini_df, how = 'left', left_on ='author', right_on = 'author')
df_final_0_1

journal    year                   author  \
0       American Economic Review  2000.0        dmitriy stolyarov   
1       American Economic Review  2000.0        eric s. rosengren   
2       American Economic Review  2000.0             avner shaked   
3       American Economic Review  2000.0        david e. wildasin   
4       American Economic Review  2000.0        thomas j. nechyba   
...                          ...     ...                      ...   
5673  Review of Economic Studies  2019.0            junehyuk jung   
5674  Review of Economic Studies  2019.0      jeong ho (john) kim   
5675  Review of Economic Studies  2019.0          eduardo morales   
5676  Review of Economic Studies  2019.0            andrés zahler   
5677  Review of Economic Studies  2019.0  eduardo souza-rodrigues   

                      name Short-Id  
0     dmitriy l. stolyarov   pst142  
1           eric rosengren    pro88  
2                      NaN      NaN  
3           david wildasin     pwi3  
4           thomas nechyba    pne28  
...                    ...      ...  
5673                   NaN      NaN  
5674                   NaN      NaN  
5675                   NaN      NaN  
5676                   NaN      NaN  
5677                   NaN      NaN  

[5678 rows x 5 columns]

In [69]:
df_final_1.columns = ["journal", "year", "author", "name", "Short-Id"]

In [70]:
df_final_1

journal    year             author  \
0        American Economic Review  2000.0    D. Gale Johnson   
2        American Economic Review  2000.0    Boyan Jovanovic   
4        American Economic Review  2000.0           Joe Peek   
5        American Economic Review  2000.0    Larry Samuelson   
6        American Economic Review  2000.0  George J. Mailath   
...                           ...     ...                ...   
13019  Review of Economic Studies  2019.0     Elisa Faraglia   
13020  Review of Economic Studies  2019.0      Albert Marcet   
13021  Review of Economic Studies  2019.0    Rigas Oikonomou   
13022  Review of Economic Studies  2019.0       Andrew Scott   
13027  Review of Economic Studies  2019.0        Gloria Sheu   

                    name Short-Id  
0        D. Gale Johnson   pjo254  
2        Boyan Jovanovic    pjo20  
4               Joe Peek    ppe90  
5        Larry Samuelson    psa80  
6      George J. Mailath     pma6  
...                  ...      ...  
13019     Elisa Faraglia   pfa164  
13020      Albert Marcet   pma173  
13021    Rigas Oikonomou     poi7  
13022       Andrew Scott    psc65  
13027        Gloria Sheu   psh447  

[7458 rows x 5 columns]

In [71]:
df_final_final = df_final_1.append(df_final_0_1)
df_final_final

journal    year                   author  \
0       American Economic Review  2000.0          D. Gale Johnson   
2       American Economic Review  2000.0          Boyan Jovanovic   
4       American Economic Review  2000.0                 Joe Peek   
5       American Economic Review  2000.0          Larry Samuelson   
6       American Economic Review  2000.0        George J. Mailath   
...                          ...     ...                      ...   
5673  Review of Economic Studies  2019.0            junehyuk jung   
5674  Review of Economic Studies  2019.0      jeong ho (john) kim   
5675  Review of Economic Studies  2019.0          eduardo morales   
5676  Review of Economic Studies  2019.0            andrés zahler   
5677  Review of Economic Studies  2019.0  eduardo souza-rodrigues   

                   name Short-Id  
0       D. Gale Johnson   pjo254  
2       Boyan Jovanovic    pjo20  
4              Joe Peek    ppe90  
5       Larry Samuelson    psa80  
6     George J. Mailath     pma6  
...                 ...      ...  
5673                NaN      NaN  
5674                NaN      NaN  
5675                NaN      NaN  
5676                NaN      NaN  
5677                NaN      NaN  

[13136 rows x 5 columns]

In [72]:
df_final_final.to_csv("autores_final_final.csv")